### ISSUE TITLE
Accuracy metric with optional num_classes argument

### ISSUE THREAD

**Issue by LarsHill** [Lightning-AI/torchmetrics#1755]
## 🚀 Feature

The fixed `num_classes` parameter in the Multiclass Accuracy calculation should be made Optional.

### Motivation

Classification problems can have a dynamic number of output classes. E.g. ordering problems, like predicting the correct word order in a shuffled sentence, depend on the number of words per sentence. Since each sentence is likely to have a different number of words, the number of "classes" differs between samples.

For example, given 2 sentences with radnomly shuffled words, the ground truth word ordering ids could look like:
Sentence 1 (6 words): 3, 5, 1, 0, 2, 4
Sentence 2 (3 words): 2, 0, 1

As a first performance proxy i want to calculate accuracy scores, by simply checking the overlap of correctly predicted positions. However, the non-optional `num_classes` argument makes it difficult to instantiate the accuracy metric in such a scenario with no fixed `num_classes`. Of course I could select a very high number of classes, hoping that I never encounter a sentence with more words, but this seems like an ugly hack.

Imho the accuracy calulation should be independent of `num_classes` in such a case. 

Is there a particular reason, why the number of classes is always needed?

A simple accuracy implementation like the following wouldn't need it for sure:

```python
def masked_accuracy(output: torch.Tensor, target: torch.Tensor, mask: torch.Tensor) -> float:
    """Compute accuracy of softmax output with mask applied over values.
    """

    with torch.no_grad():
        masked_output = torch.masked_select(output, mask)
        masked_target = torch.masked_select(target, mask)
        accuracy = masked_output.eq(masked_target).float().mean()
        return accuracy
```

I get that you handle way more cases and different averaging options, but I still feel the above case is not properly covered in your implementation...

### Pitch

Make `num_classes` optional and in cases where it is really needed, raise a custom Exception, telling the user that this argument is needed in this specific accuracy configuration.

**Comment by SkafteNicki** [comment]
Hi @LarsHill, thanks for raising this issue.
There are a multiple reason that after the classification refactor in v0.10 of torchmetrics, the `num_classes` argument has been required:
* Consistency: regardless of classification metric the interface and the arguments are the same now. While it is correct that `num_classes` are not needed in this case for accuracy, for other metrics it is always needed
* Scope: computing `MulticlassAccuracy` with anything else than `average="micro"` (which is the value you refer to) requires that the user provides `num_classes` because we need to aggregate statistics per class. Therefore, since we only could make `num_classes` optional in a single case, it seemed limited in scope.
* Verification: by forcing the user to provide the `num_classes` we can verify that data fulfils a number of constrains making sure that everything is as expected.

Finally, there is the question if such aggregation over samples with different number of classes is correct based on a theoretically point of view. If I have 3 balanced classes and random guess I have a baseline accuracy value of 33%. If I increase to 4 classes, that number becomes 25%. Thus, conclusions on metric values should always be done in relation to a base level, which I think is harder to do when that base level is dynamic.

All that said, if you are just interested in the accuracy value per batch, then I see no problem using the functional version:
```python
acc = torchmetrics.functional.classification.multiclass_accuracy(
    preds, 
    target, 
    num_classes=max(max(preds), max(target)), 
    average="micro"
)
```
where the `num_classes` argument is set dynamically.
